src/client/Microsoft.Identity.Client/Platforms/Features/SLC/KeyMaterialManager.cs

```diff
@@ -0,0 +1,370 @@
+                    _logger.Verbose(() => "[Managed Identity] Another thread created the certificate while waiting for the lock.");
+                    _isInitialized = true;
+                    return s_bindingCertificate;
+                }
+
+                // The cached certificate needs to be rotated or does not exist
+                ECDsaCng cngkey = GetCngKey();
+
+                if (cngkey != null)
+                {
+                    s_bindingCertificate = CreateCngCertificate(cngkey);
+                    _isInitialized = true;
+                    return s_bindingCertificate;
+                }
+            }
+
+            _isInitialized = false;
+            return null;
+        }
+
+        /// <summary>
+        /// Determines if a given X.509 certificate needs rotation based on a percentage threshold.
+        /// </summary>
+        /// <param name="certificate">The X.509 certificate to evaluate.</param>
+        /// <param name="rotationPercentageThreshold">The threshold percentage for considering rotation (default is 70%).</param>
+        /// <returns>
+        /// True if the certificate needs rotation, false otherwise.
+        /// </returns>
+        public static bool CertificateNeedsRotation(X509Certificate2 certificate, double rotationPercentageThreshold = 70)
+        {
+            DateTime now = DateTime.UtcNow;
+
+            // Calculate the total duration of the certificate's validity
+            TimeSpan certificateLifetime = certificate.NotAfter - certificate.NotBefore;
+
+            // Calculate how much time has passed since the certificate's issuance
+            TimeSpan elapsedTime = now - certificate.NotBefore;
+
+            // Calculate the current percentage of the certificate's lifetime that has passed
+            double percentageElapsed = (elapsedTime.TotalMilliseconds / certificateLifetime.TotalMilliseconds) * 100.0;
+
+            // Check if the percentage elapsed exceeds the rotation threshold
+            return percentageElapsed >= rotationPercentageThreshold;
+        }
+
+        /// <summary>
+        /// Initializes and retrieves the ECDsaCng key for Managed Identity.
+        /// </summary>
+        /// <returns>
+        /// The initialized ECDsaCng key if successful, otherwise null.
+        /// </returns>
+        public ECDsaCng GetCngKey()
+        {
+            _logger.Verbose(() => "[Managed Identity] Initializing Cng Key.");
+
+            // Try to get the key material from machine key
+            if (TryGetKeyMaterial(KeyProviderName, MachineKeyName, CngKeyOpenOptions.MachineKey, out ECDsaCng eCDsaCng))
+            {
+                _logger.Verbose(() => $"[Managed Identity] A machine key was found. Key Name : {MachineKeyName}. ");
+                return eCDsaCng;
+            }
+
+            // If machine key is not available, fall back to software key
+            if (TryGetKeyMaterial(KeyProviderName, SoftwareKeyName, CngKeyOpenOptions.None, out eCDsaCng))
```

Comment:
What will be the `CryptoKeyType` in this case?

Reply:
modified the logic and added comments